src/paho/mqtt/client.py

```diff
@@ -958,27 +966,29 @@ def publish(self, topic, payload=None, qos=0, retain=False):
         the length of the payload is greater than 268435455 bytes."""
         if topic is None or len(topic) == 0:
             raise ValueError('Invalid topic.')
-        if qos<0 or qos>2:
+
+        topic = topic.encode('utf-8')
+
+        if self._topic_wildcard_len_check(topic) != MQTT_ERR_SUCCESS:
+            raise ValueError('Publish topic cannot contain wildcards.')
+
+        if qos < 0 or qos > 2:
             raise ValueError('Invalid QoS level.')
-        if isinstance(payload, str) or isinstance(payload, bytearray):
-            local_payload = payload
-        elif sys.version_info[0] == 3 and isinstance(payload, bytes):
-            local_payload = bytearray(payload)
-        elif sys.version_info[0] < 3 and isinstance(payload, unicode):
+
+        if isinstance(payload, unicode):
```

Comment:
not all conversions have the same needs (and semantic), and I have already removed many duplicates. but this may be a good idea for message and will payloads.